File: 2015/day6.js
const input = require('./input/day6')

const size = 1000
const followDirections = (input, grid, instructions, reducer)=> {
  const handleInstruction = (x1, x2, y1, y2, callback) => {
    grid.forEach((row, i) => {
      if (i >= x1 && i <= x2) {
        row.forEach((elm, j) => {
          if (j >= y1 && j <= y2) {
            grid[i][j] = callback(grid[i][j])
          }
        })
      }
    })
  }
  input.split("\n").forEach(direction => {
    const [_, instruction, x1, y1, x2, y2] = direction.match(/(\D+)(\d+),(\d+) through (\d+),(\d+)/)
    handleInstruction(x1, x2, y1, y2, instructions[instruction])
  })
  return grid.reduce(reducer, 0)
}

let grid = (new Array(size)).fill(false).map(elm => (new Array(size)).fill(false))

const instructions = {
  'turn off ': () => false,
  'turn on ':  () => true,
  'toggle ':  (x) => !x
}

const reducer = (total, row) => row.filter(elm => elm).length + total
console.log(followDirections(input, grid, instructions, reducer))

let newGrid = (new Array(size)).fill(false).map(elm => (new Array(size)).fill(0))

const newInstructions = {
  'turn off ': x => x ? x - 1 : 0,
  'turn on ':  x => x + 1,
  'toggle ':  x => x + 2
}

const newReducer = (total, row) => total + row.reduce((rowTotal, elm) => rowTotal + elm, 0)
console.log(followDirections(input, newGrid, newInstructions, newReducer))

File: 2015/day6.rb
input = File.read('./input/day6.txt')

grid = {}
1000.times.each_with_object(grid) do |i, obj|
  1000.times.each_with_object(obj) do |j, obj|
    obj["#{i},#{j}"] = false
  end
end

def parse(line)
  line.match(/(\d+),(\d+) through (\d+),(\d+)/)[1..4].map(&:to_i)
end
input.split("\n").each_with_object(grid) do |line, obj|
  x1, y1, x2, y2 = parse line
  (x1..x2).each do |x|
    (y1..y2).each do |y|
      obj["#{x},#{y}"] = if line.include? 'on'
                            true
                          elsif line.include? 'off'
                            false
                          else
                            !obj["#{x},#{y}"]
                          end
    end
  end
end
puts grid.values.select{|v| v}.length

1000.times.each_with_object(grid) do |i, obj|
  1000.times.each_with_object(obj) do |j, obj|
    obj["#{i},#{j}"] = 0
  end
end

input.split("\n").each_with_object(grid) do |line, obj|
  x1, y1, x2, y2 = parse line
  (x1..x2).each do |x|
    (y1..y2).each do |y|
      key = "#{x},#{y}"
      obj[key] = if line.include? 'on'
                   obj[key] + 1
                 elsif line.include? 'off'
                   obj[key] > 0 ? obj[key] - 1 : 0
                 else
                   obj[key] + 2
                 end
    end
  end
end
puts grid.values.reduce(0){ |total, n| total + n }
